### PR TITLE
[RLlib] Add `worker` arg (optional) to `policy_mapping_fn`.

### PR DESCRIPTION
**Pull request by sven1977**
Adds `worker` (RolloutWorker) arg (optional) to `policy_mapping_fn`.
This is useful if policy mapping and agent matching (e.g. in adversarial games) depends on the worker's index.

E.g.:
```
def policy_mapping_fn(agent_id, episode, worker, **kwargs):
    if worker.worker_index == 1:
        return "pol1"
    elif worker.worker_index == 2:
        return "pol2"
...
```

The new signature is:
`policy_mapping_fn(agent_id, episode, worker, **kwargs)`
This is fully backward compatible b/c:
1) All policy_mapping_fn should accept **kwargs and calls are always made with the new worker arg being named (`policy_mapping_fn(agent_id, episode, worker=worker)`).
2) Even if policy_mapping_fn still use the old signature (`policy_mapping_fn(agent_id)`), RLlib will catch the exception and try the old signature (w/o episode and worker args).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/evaluation/episode.py [hunk 1]
```diff
@@ -120,17 +140,17 @@ def policy_for(self, agent_id: AgentID = _DUMMY_AGENT_ID) -> PolicyID:
             # New signature should be: (agent_id, episode, **kwargs)
```
Comment (sven1977) [inline rllib/evaluation/episode.py:140]:
```suggestion
            # New signature should be: (agent_id, episode, worker, **kwargs)
```

**Review by avnishn — COMMENTED** [review]
So if I understand correctly, in this PR you've added a parameter to the policy mapping function, worker, that can be used in mapping policies to agents. The example you gave uses the worker's ID in order to help with the mapping. Is there a reason why you would need other information from the worker, other than its id? 

If you only need the worker's id, then wouldn't it be better to just pass the worker ID as a kwarg?

**Comment by sven1977** [comment]
Yeah, good question. In this particular case, the user had asked to do mapping based on the worker index, but in general, I think it would be good to have access to the entire RolloutWorker as it contains lots of other information that may be useful for figuring out the agent->policy matching. For example:
```
worker.num_workers  # number of workers in total
worker.policy_map.keys()  # list of all available policies
worker.async_env  # actual (vectorized) environment object
...
```

**Comment by avnishn** [comment]
> Yeah, good question. In this particular case, the user had asked to do mapping based on the worker index, but in general, I think it would be good to have access to the entire RolloutWorker as it contains lots of other information that may be useful for figuring out the agent->policy matching. For example:
> 
> ```
> worker.num_workers  # number of workers in total
> worker.policy_map.keys()  # list of all available policies
> worker.async_env  # actual (vectorized) environment object
> ...
> ```

Gotcha, lgtm

**Review by avnishn — APPROVED** [review]

**Pull request merged**